blockprod/src/detail/mod.rs

```diff
@@ -392,302 +349,398 @@ impl BlockProduction {
+            (RequiredConsensus::IgnoreConsensus, GenerateBlockInputData::PoS(_)) => {
+                Err(BlockProductionError::PoSInputDataProvidedWhenIgnoringConsensus)?
+            }
+            (RequiredConsensus::IgnoreConsensus, GenerateBlockInputData::PoW(_)) => {
+                Err(BlockProductionError::PoWInputDataProvidedWhenIgnoringConsensus)?
+            }
+        }
+    }
 
+    async fn produce_block_pos(
+        &self,
+        helper: Helper,
+        pos_status: PoSStatus,
+        input_data: PoSGenerateBlockInputData,
+        transactions: Vec<SignedTransaction>,
+        transaction_ids: Vec<Id<Transaction>>,
+        packing_strategy: PackingStrategy,
+    ) -> Result<(Block, oneshot::Receiver<usize>), BlockProductionError> {
         // Unlike Proof of Work, which can vary any header field when
         // searching for a valid block, Proof of Stake can only vary
         // the header timestamp. Its search space starts at the
         // previous block's timestamp + 1 second, and ends at the
         // current timestamp + some distance in time defined by the
         // blockchain.
-        //
-        // This variable keeps track of the last timestamp that was
-        // attempted, and during Proof of Stake, will prevent
-        // searching over the same search space, across multiple
-        // calls, given the same tip
-        let last_timestamp_seconds_used = {
-            let tip_timestamp = cmp::max(
-                previous_last_used_block_timestamp.unwrap_or(BlockTimestamp::from_int_seconds(0)),
-                tip_at_start.block_timestamp(),
-            );
+        let starting_timestamp = helper.starting_timestamp();
+        let max_timestamp = helper.max_timestamp();
+
+        // Note/TODO: the "vrf_data" part of consensus_data is useless here, because it'll be unconditionally
+        // overwritten below. Perhaps we could introduce an intermediate "PartialPoSData" struct that wouldn't
+        // contain the vrf data, so that it could be produced by pull_consensus_data_pos instead.
+        let (consensus_data, block_reward, finalize_block_data) = self
+            .pull_consensus_data_pos(
+                input_data,
+                pos_status.clone(),
+                helper.tip_block_index(),
+                starting_timestamp,
+            )
+            .await?;
 
-            let tip_plus_one = tip_timestamp
-                .add_int_seconds(1)
-                .ok_or(ConsensusCreationError::TimestampOverflow(tip_timestamp, 1))?;
+        log::debug!(
+            "Searching for a valid block ({}..={}), pool_id: {}",
+            starting_timestamp,
+            max_timestamp,
+            Address::new(&self.chain_config, *consensus_data.stake_pool_id())
+                .expect("Pool id to address cannot fail")
+        );
 
-            Arc::new(AcqRelAtomicU64::new(tip_plus_one.as_int_seconds()))
-        };
+        let search_start_time = std::time::Instant::now();
```

Comment:
Please be aware that using Instant circumvents the mocked time mechanics for regtest.